plutus-core/plutus-core/src/PlutusCore/Default/Builtins.hs

```diff
@@ -132,6 +130,323 @@ nonZeroArg :: (Integer -> Integer -> Integer) -> Integer -> Integer -> Evaluatio
+        makeBuiltinMeaning
+            BS.length
+            <costingFunction>
+
+directly (where @BS.length :: BS.ByteString -> Int@).
+
+Note however that while it's always safe to convert an 'Int' to an 'Integer', doing the opposite is
+not safe in general, because an 'Integer' may not fit into the range of 'Int'. For this reason
+
+    YOU MUST NEVER USE 'fromIntegral' AND SIMILAR FUNCTIONS THAT CAN SILENTLY UNDER- OR OVERFLOW
+    WHEN DEFINING A BUILT-IN FUNCTION
+
+For example defining a builtin that takes an 'Integer' and converts it to an 'Int' using
+'fromIntegral' is not allowed under any circumstances and can be a huge vulnerability.
+
+It's completely fine to define a builtin that takes an 'Int' directly, though. How so? That's due
+to the fact that the builtin application machinery checks that an 'Integer' is in the bounds of
+'Int' before doing the conversion. If the bounds check succeeds, then the 'Integer' gets converted
+to the corresponding 'Int', and if it doesn't, then the builtin application fails.
+
+For the list of types that can be converted to/from built-in ones look into the file with the
+default universe. If you need to add a new such type, just copy-paste what's done for an existing
+one and adjust.
+
+Speaking of builtin application failing:
+
+5. A built-in function can fail. Whenever a builtin fails, evaluation of the whole program fails.
+There's a number of ways a builtin can fail:
+
+- as we've just seen a type conversion can fail due to an unsuccessful bounds check
+- if the builtin expects, say, a 'Text' argument, but gets fed an 'Integer' argument
+- if the builtin expects any constant, but gets fed a non-constant
+- if its denotation runs in the 'EvaluationResult' and an 'EvaluationFailure' gets returned
+
+Most of these are not a concern to the user defining a built-in function (conversions are handled
+within the builtin application machinery, type mismatches are on the type checker and the person
+writing the program etc), however explicitly returning 'EvaluationFailure' from a builtin is
+something that happens commonly.
+
+One simple example is a monomorphic function matching on a certain constructor and failing in all
+other cases:
+
+    toBuiltinMeaning UnIData =
+        makeBuiltinMeaning
+            (\case
+                I i -> EvaluationSuccess i
+                _   -> EvaluationFailure)
+            <costingFunction>
+
+The inferred type of the denotation is
+
+    Data -> EvaluationResult Integer
+
+and the Plutus type of the builtin is
+
+    data -> integer
+
+because the error effect is implicit in Plutus.
+
+Returning @EvaluationResult a@ for a type variable @a@ is also fine, i.e. it doesn't matter whether
+the denotation is monomorphic or polymorphic w.r.t. failing.
+
+But note that
+
+    'EvaluationResult' MUST BE EXPLICIYLY USED FOR ANY FAILING BUILTIN AND THROWING AN EXCEPTION
+    VIA 'error' OR 'throw' OR ELSE IS NOT ALLOWED AND CAN BE A HUGE VULNERABILITY. MAKE SURE THAT
```

Comment:
OR ELSE?

Reply:
`errorWithoutStackTrace`, `throwTo`, you name it.